File: mediator/python/chat_room_mediator.py
# Created by Sezer BOZKIR<admin@sezerbozkir.com> at 21.12.2021
from abstract_chat_room_mediator import AbstractChatRoomMediator
from chat_user import ChatUser


class ChatRoomMediator(AbstractChatRoomMediator):
    """
    Mediator (IChatRoomMediator) arayüzünü uygular.
    Colleague (User) nesneleri arasındaki iletişimi koordine eder.
    UML diyagramındaki ConcreteMediator yapısına denk gelmektedir.
    """

    def __init__(self):
        self._user_dictionary = {}

    def send_message(self, message: str, user_id: int):
        user: ChatUser = self._user_dictionary[user_id]
        user.receive_message(message)

    def add_user_in_room(self, user: ChatUser):
        self._user_dictionary[user.u_id] = user

File: mediator/go/chat-room-mediator.go
package main

// Mediator (IChatRoomMediator) arayüzünü uygular.
// Colleague (User) nesneleri arasındaki iletişimi koordine eder.
// UML diyagramındaki ConcreteMediator yapısına denk gelmektedir.
type chatRoomMediator struct {
	users map[int]*User
}

func NewChatRoomMediator() *chatRoomMediator {
	return &chatRoomMediator{
		users: make(map[int]*User),
	}
}

func (m *chatRoomMediator) AddUserInRoom(user *User) {
	m.users[user.ID] = user
}

func (m chatRoomMediator) SendMessage(message string, userID int) {
	user, k := m.users[userID]
	if !k {
		return
	}
	user.ReceiveMessage(message)
}
